ocaml/xapi/helpers.ml

```diff
@@ -1234,3 +1234,108 @@ let host_supports_hvm ~__context host =
+      * TODO: don't generate certificates in bash *)
+    val gen_pem_if_not_exist : __context:Context.t -> unit
+  end = struct
+    let xapi_stunnel_conf = "/etc/stunnel/xapi.conf"
+    let m = Mutex.create ()
+    let current_accept = ref None
+
+    let rewrite_xapi_ssl_config_file ~accept =
+      let idle_timeout = match Sys.getenv_opt "STUNNEL_IDLE_TIMEOUT" with
+                         | None -> "; no idle timeout"
+                         | Some x -> Printf.sprintf "TIMEOUTidle = %s" x
+      in
+      let conf_contents = String.concat "\n" [
+          "; autogenerated by helpers.ml"
+        ; "fips = no"
+        ; "pid = /var/run/xapissl.pid"
+        ; "socket = r:TCP_NODELAY=1"
+        ; "socket = a:TCP_NODELAY=1"
+        ; "socket = l:TCP_NODELAY=1"
+        ; "socket = r:SO_KEEPALIVE=1"
+        ; "socket = a:SO_KEEPALIVE=1"
+        ; idle_timeout
+        ; "debug = authpriv.5"
+        ; ""
+        ; "[xapi]"
+        ; Printf.sprintf "accept = %s" accept
+        ; "connect = 80"
+        ; Printf.sprintf "cert = %s" cert
+        ; Printf.sprintf "ciphers = %s" Xcp_const.good_ciphersuites
+        ; "curve = secp384r1"
+        ; "TIMEOUTclose = 1"
+        ; "options = CIPHER_SERVER_PREFERENCE"
+        ; "sslVersion = TLSv1.2"
+        ]
+      in
+      let len = String.length conf_contents in
+      Stdext.Unixext.atomic_write_to_file xapi_stunnel_conf 0o0600
+        (fun fd -> Unix.single_write_substring fd conf_contents 0 len |> ignore)
+
+    let update_config ~accept =
+      Mutex.execute m (fun () ->
+        match !current_accept with
+        | Some current_accept when current_accept = accept -> ()
+        | None | Some _ -> begin
+          current_accept := Some accept;
+          rewrite_xapi_ssl_config_file ~accept
+        end
+      )
+
+    let gen_pem_if_not_exist ~__context =
+      Mutex.execute m (fun () ->
+        if not (Sys.file_exists cert) then begin
+          let hostname = Unix.gethostname () in
+          let cn =
+            if Astring.String.is_prefix ~affix:"localhost" hostname then
+              get_management_ip_addr ~__context
+            else if Astring.String.is_infix ~affix:"." hostname then
+              None
+            else
+              get_management_ip_addr ~__context
+          in
+          call_script "/opt/xensource/libexec/generate_ssl_cert" [cert; Opt.default hostname cn] |> ignore
+        end
+      )
+
+  end
+
+  let systemctl cmd = call_script "/usr/bin/systemctl" [cmd; "stunnel@xapi"]
+  let systemctl_ cmd = systemctl cmd |> ignore
+
+  let is_enabled () =
+    systemctl "is-enabled" |>
+    Astring.String.trim |>
+    function
+    | "enabled"  -> true
+    | "disabled" -> false
+    | _          -> D.error "Stunnel.is_enabled: unable to discern if stunnel@xapi is enabled. assuming that it is disabled"; false
+
+  let restart ~__context ~accept =
+    try
+      File.update_config ~accept;
+      File.gen_pem_if_not_exist ~__context;
```

Comment:
Previously this was done on every start of the daemon (on `start ()`)
While I don't like the idea of having logic for creating certs run every time stunnel is started, I wonder if the replacement is enough: stunnel could be started up without xapi being involved, missing the creation of the certificate.

I don't think it's happening right now, but with all the changes on the service management and firstboot some breakage could slip through.

Could you check this suspicion so we're confident this change doesn't add brittleness?

Reply:
You could add another systemd service file, and make stunnel@xapi depend on it (it should be possible to do this from your gencert.service by adding a WantedBy or something similar).
Then systemd would guarantee us that it is available.